internal/pkg/flink/internal/controller/input_controller.go

```diff
@@ -32,258 +22,85 @@ type InputController struct {
-			cancelListenToUserInput()
-			outputErr(err.Error())
-			if !c.isSessionValid(err) {
-				c.appController.ExitApplication()
-				return
-			}
-			continue
-		}
-		processedStatement.PrintStatusDetail()
-
-		processedStatement, err = c.store.FetchStatementResults(*processedStatement)
-		cancelListenToUserInput()
-		if err != nil {
-			outputErr(err.Error())
-			continue
-		}
+const defaultWindowSize = 100
 
-		// decide if we want to display results using TView or just a plain table
-		if shouldUseTView(*processedStatement) {
-			c.table.Init(*processedStatement)
-			return
-		}
-
-		c.printResultToSTDOUT(processedStatement)
-		// This was used to delete statements after their execution to save system resources, which should not be
-		// an issue anymore. We don't want to remove it completely just yet, but will disable it by default for now.
-		// TODO: remove this completely once we are sure we won't need it in the future
-		if config.ShouldCleanupStatements && !processedStatement.IsLocalStatement && processedStatement.Status != types.RUNNING {
-			go c.store.DeleteStatement(processedStatement.StatementName)
-		}
+func NewInputController(a types.ApplicationControllerInterface, history *history.History) types.InputControllerInterface {
+	inputController := &InputController{
+		History:         history,
+		InitialBuffer:   "",
+		appController:   a,
+		smartCompletion: true,
+		shouldExit:      false,
+		reverseISearch:  reverseisearch.NewReverseISearch(),
 	}
+	inputController.prompt = inputController.Prompt()
+	return inputController
 }
 
-func (c *InputController) listenToUserInput(in prompt.ConsoleParser, cancelFunc context.CancelFunc) context.CancelFunc {
-	ctx, cancelListenToUserInput := context.WithCancel(context.Background())
-	go func() {
-		for {
-			select {
-			case <-ctx.Done():
-				return
-			default:
-				if b, err := in.Read(); err == nil && len(b) > 0 {
-					pressedKey := prompt.Key(b[0])
-
-					switch pressedKey {
-					case prompt.ControlC:
-						fallthrough
-					case prompt.ControlD:
-						fallthrough
-					case prompt.ControlQ:
-						fallthrough
-					case prompt.Escape:
-						// esc
-						cancelFunc()
-						return
-					}
-				}
-			}
-			time.Sleep(10 * time.Millisecond)
-		}
-	}()
-	return cancelListenToUserInput
-}
-
-func (c *InputController) isSessionValid(err *types.StatementError) bool {
-	// exit application if user needs to authenticate again
-	if err != nil && err.HttpResponseCode == http.StatusUnauthorized {
-		return false
+func (c *InputController) GetUserInput() string {
+	// if the initial buffer is not empty, we insert the text and reset the InitialBuffer
+	if c.InitialBuffer != "" {
+		c.prompt.Buffer().InsertText(c.InitialBuffer, false, true)
+		c.InitialBuffer = ""
 	}
-	return true
+	return c.prompt.Input()
 }
 
-func renderMsgAndStatus(processedStatement *types.ProcessedStatement) {
-	if processedStatement == nil {
-		return
+func (c *InputController) IsSpecialInput(userInput string) bool {
```

Comment:
Add some comments here because these are kinda of hacks and might be not so intuitive to understand

Reply:
Ah right, especially the empty string return from go-prompt right?